tests/rbd_mirror/test_image_operations_on_snap_mirroring.py

```diff
@@ -0,0 +1,174 @@
+from ceph.rbd.initial_config import initial_mirror_config
+from ceph.rbd.utils import getdict
+from ceph.rbd.workflows.cleanup import cleanup
+from utility.log import Log
+
+log = Log(__name__)
+
+
+def test_image_remove_resize(rbd, pool, image, image_size, cluster_type):
+    """
+    Perform image remove and resize and verify the output based on
+    whether its primary or secondary cluster
+    """
+    size_num = int(image_size[:-1])
+    expand_size = str(2 * size_num) + "G"
+    resize_config = {"image-spec": f"{pool}/{image}", "size": expand_size}
+    out, err = rbd.resize(**resize_config)
+    if cluster_type == "primary" and "100% complete" not in out + err:
+        log.error(
+            f"Image {image} expand failed in {cluster_type} cluster with error {out} {err}"
+        )
+        return 1
+    elif cluster_type == "secondary" and "Read-only file system" not in out + err:
+        log.error(
+            f"Image {image} expand did not fail as expected in {cluster_type} cluster"
+        )
+        return 1
+
+    resize_config["size"] = image_size
+    resize_config.update({"allow-shrink": True})
+    out, err = rbd.resize(**resize_config)
+    if cluster_type == "primary" and "100% complete" not in out + err:
+        log.error(
+            f"Image {image} shrink failed in {cluster_type} cluster with error {out} {err}"
+        )
+        return 1
+    elif (
+        cluster_type == "secondary"
+        and "new size is equal to original size" not in out + err
+    ):
+        log.error(
+            f"Image {image} shrink did not fail as expected in {cluster_type} cluster"
+        )
+        return 1
+
+    remove_config = {"image-spec": f"{pool}/{image}"}
+    out, err = rbd.rm(**remove_config)
+    if cluster_type == "primary" and "100% complete" not in out + err:
+        log.error(
+            f"Image {image} remove failed in {cluster_type} cluster with error {out} {err}"
+        )
+        return 1
+    elif (
+        cluster_type == "secondary" and "mirrored image is not primary" not in out + err
+    ):
+        log.error(
+            f"Image {image} remove did not fail as expected in {cluster_type} cluster"
+        )
+        return 1
+    return 0
+
+
+def test_image_operations(**kw):
+    """
+    Test image operations like expand/shrink/remove in primary
+    and secondary clusters
+    """
+    for pool_type in kw.get("pool_types"):
+        config = kw.get("config", {}).get(pool_type)
+        multi_pool_config = getdict(config)
+        for pool, pool_config in multi_pool_config.items():
+            multi_image_config = getdict(pool_config)
+            for image, image_config in multi_image_config.items():
+                image_size = image_config.get("size")
+
+                log.info(
```

Comment:
we can iterate line 76 to 110 inside for loop for `primary` and `secondary` for code optimization.

Reply:
Done.